src/classes/BroadPhases/BasicPooledBroadphase.js

```diff
@@ -0,0 +1,201 @@
+ */
+Goblin.BasicPooledBroadphase.prototype.getDynamicBodies = function() {
+    return this.dynamic_bodies;
+};
+
+/**
+ * Updates body's collision layer
+ *
+ * @method updateObjectLayer
+ * @param rigid_body {Goblin.RigidBody} Rigid body to update
+ * @param new_layer  {Number} New layer that is about to be set
+ */
+Goblin.BasicBroadphase.prototype.updateObjectLayer = function ( rigid_body, new_layer ) {
+    if ( rigid_body._layer !== null && this._layers[ rigid_body._layer ] ) {
+        this._removeBodyFrom( rigid_body, this._layers[ rigid_body._layer ] );
+    }
+
+    if ( new_layer !== null ) {
+        this._layers[ new_layer ].push( rigid_body );
+    }
+};
+
+/**
+ * Adds a body to the broadphase for contact checking
+ *
+ * @method addBody
+ * @param body {RigidBody} body to add to the broadphase contact checking
+ */
+Goblin.BasicPooledBroadphase.prototype.addBody = function( body ) {
+    Goblin.BasicBroadphase.prototype.addBody.call( this, body );
+    
+    if ( body.static ) {
+        this.static_bodies.push( body );
+    } else {
+        this.dynamic_bodies.push( body );
+    }
+
+    if ( body._layer !== null ) {
+        this._layers[ body._layer ].push( body );
+    }
+};
+
+/**
+ * Checks if a ray segment intersects with objects in the world
+ *
+ * @method rayIntersect
+ * @param start         {vec3}      Start point of the segment
+ * @param end           {vec3}      End point of the segment
+ * @param limit         {Number}    Limit the amount of intersections (i.e. 1)
+ * @param layerMask     {Number}    The bitmask of layers to check
+ * @return {Array<RayIntersection>} an unsorted array of intersections
+ */
+Goblin.BasicPooledBroadphase.prototype.rayIntersect = (function () {
+    // FIXME EN-77 should eliminate the below
+    var _start = new Goblin.Vector3();
+    var _end = new Goblin.Vector3();
+
+    return function( start, end, limit, layerMask ) {
+        // copy vector values over to allow for duck typing
+        _start.copy( start );
+        _end.copy( end );
+
+        var intersections = [];
+
+        for ( var i = 0; i < this._layers.length; i++ ) {
+            var objects = this._layers[ i ];
+
+            if ( layerMask && ( layerMask & (1 << i) ) === 0 ) {
+                continue;
+            }
+
+            for ( var j = 0; j < objects.length; j++ ) {
+                var body = objects[ j ];
+
+                // first test AABB intersection (~ broad phase)
+                if ( body.aabb.testRayIntersect( _start, _end ) ) {
+                    // if AABB intersects, as the body about inner intersections
+                    body.rayIntersect( _start, _end, limit, intersections );
+                }
+
+                if ( limit && ( intersections.length >= limit ) ) {
+                    return intersections.slice( 0, limit );
+                }
+            }
+        }
+
+        intersections.sort( function ( a, b ) {
```

Comment:
Comparison function can be declared in outer scope, just to keep memory

Reply:
Let's postpone this up to raycast refactoring.